File: scripts/src/create-pool.ts
import * as anchor from '@coral-xyz/anchor'
import { TransactionBuilder } from '@orca-so/common-sdk'
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
  getAssociatedTokenAddressSync,
} from '@solana/spl-token'
import {
  Keypair,
  PublicKey,
  SYSVAR_RENT_PUBKEY,
  SystemProgram,
} from '@solana/web3.js'

import { getConstantParams } from './params'
import { createTransactionChained } from './utils/txix'

async function main() {
  const {
    program,
    programId,
    connection,
    wallet,
    feeRate,
    tickSpacing,
    tokenMintA,
    tokenMintB,
    tokenOracleA,
    tokenOracleB,
    cladKey,
    initPrice,
    initSqrtPrice,
  } = await getConstantParams()

  const tokenMintAKey = tokenMintA.address
  const tokenMintBKey = tokenMintB.address

  console.log('Init B/A Price: ', initPrice.toString())

  //
  // Create Clad account
  //

  const initializeCladParams = {
    permissions: {
      allowSwap: true,
      allowAddLiquidity: true,
      allowRemoveLiquidity: true,
      allowOpenPosition: true,
      allowClosePosition: true,
    },
    protocolFeeRate: 0,
  }
  const initializeCladAccounts = {
    admin: wallet.publicKey,
    clad: cladKey,
    systemProgram: SystemProgram.programId,
    rent: SYSVAR_RENT_PUBKEY,
  }

  const txInitializeClad = new TransactionBuilder(
    connection,
    wallet
  ).addInstruction({
    instructions: [
      program.instruction.initializeClad(initializeCladParams, {
        accounts: initializeCladAccounts,
      }),
    ],
    cleanupInstructions: [],
    signers: [],
  })

  const txIdInitializeClad = await txInitializeClad.buildAndExecute()
  console.log('Initialized Clad: ', txIdInitializeClad)

  //
  // Create Pool
  //

  const globalpoolSeeds = [
    Buffer.from('globalpool'),
    tokenMintAKey.toBuffer(),
    tokenMintBKey.toBuffer(),
    new anchor.BN(feeRate).toArrayLike(Buffer, 'le', 2),
    new anchor.BN(tickSpacing).toArrayLike(Buffer, 'le', 2),
  ]

  const [globalpoolKey, globalpoolBump] = PublicKey.findProgramAddressSync(
    globalpoolSeeds,
    programId
  )

  const tokenVaultA = getAssociatedTokenAddressSync(
    tokenMintAKey,
    globalpoolKey,
    true
  )
  const tokenVaultB = getAssociatedTokenAddressSync(
    tokenMintBKey,
    globalpoolKey,
    true
  )
  // const tokenVaultA = await createAndMintToAssociatedTokenAccount(
  //   provider,
  //   tokenMintA,
  //   0,
  //   globalpoolKey,
  // )
  // const tokenVaultB = await createAndMintToAssociatedTokenAccount(
  //   provider,
  //   tokenMintB,
  //   0,
  //   globalpoolKey,
  // )

  const createPoolParams = {
    feeRate,
    tickSpacing,
    initialSqrtPrice: initSqrtPrice,
  }
  const createPoolAccounts = {
    funder: wallet.publicKey,
    clad: cladKey,
    globalpool: globalpoolKey,
    tokenMintA: tokenMintA.address,
    tokenMintB: tokenMintB.address,
    tokenVaultA: tokenVaultA,
    tokenVaultB: tokenVaultB,
    tokenPriceFeedA: tokenOracleA,
    tokenPriceFeedB: tokenOracleB,
    // sys
    tokenProgram: TOKEN_PROGRAM_ID,
    associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
    systemProgram: SystemProgram.programId,
    rent: SYSVAR_RENT_PUBKEY,
  }

  const txIdCreatePool = await createTransactionChained(
    connection,
    wallet,
    program.instruction.createPool(createPoolParams, {
      accounts: createPoolAccounts,
    }),
    []
  ).buildAndExecute()

  console.log('Created Pool: ', txIdCreatePool)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: programs/clad/src/instructions/create_pool.rs
use {
    crate::state::*,
    anchor_lang::prelude::*,
    anchor_spl::token::{self, Mint, Token, TokenAccount},
};

#[derive(Accounts)]
#[instruction(params: CreatePoolParams)]
pub struct CreatePool<'info> {
    #[account(mut)]
    pub funder: Signer<'info>,

    #[account(
        seeds = [b"clad"],
        bump = clad.clad_bump,
    )]
    pub clad: Box<Account<'info, Clad>>,

    #[account(
        init,
        seeds = [
            b"globalpool".as_ref(),
            token_mint_a.key().as_ref(),
            token_mint_b.key().as_ref(),
            params.fee_rate.to_le_bytes().as_ref(),
            params.tick_spacing.to_le_bytes().as_ref(),
        ],
        bump,
        payer = funder,
        space = Globalpool::LEN
    )]
    pub globalpool: Box<Account<'info, Globalpool>>,

    pub token_mint_a: Account<'info, Mint>,

    pub token_mint_b: Account<'info, Mint>,

    #[account(
        init,
        payer = funder,
        token::mint = token_mint_a,
        token::authority = globalpool // transfer_authority
    )]
    pub token_vault_a: Box<Account<'info, TokenAccount>>,

    #[account(
        init,
        payer = funder,
        token::mint = token_mint_b,
        token::authority = globalpool // transfer_authority
    )]
    pub token_vault_b: Box<Account<'info, TokenAccount>>,

    #[account(address = token::ID)]
    pub token_program: Program<'info, Token>,

    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct CreatePoolParams {
    pub fee_rate: u16,
    pub tick_spacing: u16,
    pub initial_sqrt_price: u128,
}

pub fn create_pool(ctx: Context<CreatePool>, params: &CreatePoolParams) -> Result<()> {
    let clad = &ctx.accounts.clad;

    let globalpool = &mut ctx.accounts.globalpool;
    let globalpool_bump = *ctx
        .bumps
        .get("globalpool")
        .ok_or(ProgramError::InvalidSeeds)?;

    Ok(globalpool.initialize(
        globalpool_bump,
        params.tick_spacing,
        params.initial_sqrt_price,
        params.fee_rate,
        clad.protocol_fee_rate,
        ctx.accounts.funder.key(),
        ctx.accounts.token_mint_a.key(),
        ctx.accounts.token_vault_a.key(),
        ctx.accounts.token_mint_b.key(),
        ctx.accounts.token_vault_b.key(),
    )?)
}
